icechunk-python/tests/test_stateful_parallel_sessions.py

```diff
@@ -0,0 +1,232 @@
+import zarr
+import zarr.testing.strategies as zrst
+from icechunk.distributed import merge_sessions
+from zarr.abc.store import Store
+from zarr.testing.stateful import SyncStoreWrapper
+
+zarr.config.set({"array.write_empty_chunks": True})
+
+
+def simple_dtypes() -> st.SearchStrategy[np.dtype[Any]]:
+    return npst.integer_dtypes(endianness="=") | npst.floating_dtypes(endianness="=")
+
+
+class SerialParallelStateMachine(RuleBasedStateMachine):
+    """
+    This stateful test asserts that two stores :
+    1. one on which all actions are executed in serial
+    2. one on which those same actions may be executed on the parent session,
+       or on forks. Importantly, forks may be created with a 'dirty' state.
+
+    To model this we use the same repo with two branches.
+    """
+
+    def __init__(self) -> None:
+        super().__init__()
+
+        self.storage = ic.local_filesystem_storage(
+            f"tmp/icechunk_parallel_stateful/{str(datetime.datetime.now()).split(' ')[-1]}"
+        )
+        self.repo = ic.Repository.create(self.storage)
+        self.repo.create_branch("parallel", self.repo.lookup_branch("main"))
+
+        # TODO: should this just be Zarr memory store instead?
+        #       are version control ops on the serial store useful?
+        self.serial = self.repo.writable_session("main")
+        self.parallel = self.repo.writable_session("parallel")
+
+        self.fork1: ic.Session | None = None
+        self.fork2: ic.Session | None = None
+
+        self.has_changes = False
+        self.all_arrays: set[str] = set()
+
+    def has_forks(self) -> bool:
+        return self.fork1 is not None and self.fork2 is not None
+
+    @precondition(lambda self: not self.has_forks())
+    @rule(
+        data=st.data(),
+        name=zrst.node_names,
+        array_and_chunks=zrst.np_array_and_chunks(
+            arrays=npst.arrays(simple_dtypes(), npst.array_shapes())
+        ),
+    )
+    def add_array(
+        self,
+        data: st.DataObject,
+        name: str,
+        array_and_chunks: tuple[np.ndarray[Any, Any], tuple[int, ...]],
+    ) -> None:
+        array, _ = array_and_chunks
+        # TODO: support size-0 arrays GH392
+        assume(array.size > 0)
+        array, chunks = array_and_chunks
+        fill_value = data.draw(npst.from_dtype(array.dtype))
+        assume(name not in self.all_arrays)
+        note(f"Adding array:  path='{name}'  shape={array.shape}  chunks={chunks}")
+        for store in [self.serial.store, self.parallel.store]:
+            zarr.array(
+                array,
+                chunks=chunks,
+                path=name,
+                store=store,
+                fill_value=fill_value,
+                zarr_format=3,
+                dimension_names=None,
+            )
+        self.all_arrays.add(name)
+
+    @precondition(lambda self: bool(self.all_arrays))
+    @rule(data=st.data())
+    def write_chunk(self, data: st.DataObject) -> None:
```

Comment:
This is problematic. We could draw the same path on two draws which would model the case where parallel writers write to the same chunk and behaviour is undefined. I guess I can filter on chunk paths not in `self.serial.status().updated_chunks[array]`